pandas/core/ops.py

```diff
@@ -1058,69 +1058,76 @@ def _bool_method_SERIES(cls, op, special):
-                else:
-                    x = _ensure_object(x)
-                    y = _ensure_object(y)
-                    result = lib.vec_binop(x, y, op)
+            assert not isinstance(y, (list, ABCSeries, pd.Index))
+            if isinstance(y, np.ndarray):
+                # This next assertion is here because there used to be
+                # a branch that specifically caught this case.
+                assert not (is_bool_dtype(x) and is_bool_dtype(y))
+                x = _ensure_object(x)
+                y = _ensure_object(y)
+                result = lib.vec_binop(x, y, op)
             else:
                 # let null fall thru
+                assert is_scalar(y)
                 if not isna(y):
                     y = bool(y)
                 try:
                     result = lib.scalar_binop(x, y, op)
                 except:
-                    msg = ("cannot compare a dtyped [{dtype}] array "
-                           "with a scalar of type [{type}]"
-                           ).format(dtype=x.dtype, type=type(y).__name__)
-                    raise TypeError(msg)
+                    raise TypeError("cannot compare a dtyped [{dtype}] array "
+                                    "with a scalar of type [{typ}]"
+                                    .format(dtype=x.dtype,
+                                            typ=type(y).__name__))
 
         return result
 
+    fill_int = lambda x: x.fillna(0)
+    fill_bool = lambda x: x.fillna(False).astype(bool)
+
     def wrapper(self, other):
         is_self_int_dtype = is_integer_dtype(self.dtype)
 
-        fill_int = lambda x: x.fillna(0)
-        fill_bool = lambda x: x.fillna(False).astype(bool)
-
         self, other = _align_method_SERIES(self, other, align_asobject=True)
 
+        res_name = _get_series_op_result_name(self, other)
+
         if isinstance(other, ABCDataFrame):
             # Defer to DataFrame implementation; fail early
             return NotImplemented
 
-        elif isinstance(other, ABCSeries):
-            name = com._maybe_match_name(self, other)
+        elif isinstance(other, (ABCSeries, pd.Index)):
             is_other_int_dtype = is_integer_dtype(other.dtype)
             other = fill_int(other) if is_other_int_dtype else fill_bool(other)
-
-            filler = (fill_int if is_self_int_dtype and is_other_int_dtype
-                      else fill_bool)
-
-            res_values = na_op(self.values, other.values)
-            unfilled = self._constructor(res_values,
-                                         index=self.index, name=name)
-            return filler(unfilled)
+            ovalues = other.values
 
         else:
             # scalars, list, tuple, np.array
-            filler = (fill_int if is_self_int_dtype and
-                      is_integer_dtype(np.asarray(other)) else fill_bool)
+            is_other_int_dtype = is_integer_dtype(np.asarray(other))
+            if isinstance(other, list):
```

Comment:
ATM lists are handled inside na_op on 1066-1067.  This is just moving the same two lines up the call stack (and outside of an exception-handling block)

Reply:
the comment also includes tuple and np.array, should not this check for also for tuple?

Reply:
I guess it depends on if we want to treat tuples as "an array the user forgot to wrap in an array".  For lists we do that pretty consistently (e.g. on line 1034 and 1172 in core.ops), but not so much for tuples.  I don't have a strong opinion as to what the ideal behavior is, but would prefer to address it separately from this cleanup/testing PR.  (it could go in the same Issue as the TODO comment just below this isinstance check)

Reply:
this should be is_list_like